torch/csrc/jit/passes/subgraph_matcher.cpp

```diff
@@ -0,0 +1,269 @@
+      auto output = outputs[i];
+      if (match->get(match_output) == output) {
+        std::stringstream ss;
+        ss << "[info] already matched output value " << match_output << " to " << output;
+        log(ss.str());
+        continue;
+      }
+      REQ(matchValues(match_output, output));
+      match->set(match_output, output);
+      if (match_output->node() && output->node()) {
+        value_frontier->emplace_back(std::make_pair(match_output, output));
+      }
+    }
+  }
+
+  return true;
+}
+
+bool MatchIterator::matchValueFrontier(
+    const ValueFrontier& value_frontier,
+    MatchIterator::Match* match,
+    NodeFrontier* node_frontier) {
+  node_frontier->clear();
+  if (!value_frontier.size()) {
+    return true;
+  }
+  for (const auto value_pair : value_frontier) {
+    auto match_node = value_pair.first->node();
+    auto node = value_pair.second->node();
+    do {
+      if (match->get(match_node) == node) {
+        std::stringstream ss;
+        ss << "[info] already matched node " << match_node << " to " << node;
+        log(ss.str());
+        continue;
+      }
+      if (match_node->kind() == prim::Param) {
+        continue;
+      }
+      REQ(node->owningBlock() == current_block_);
+      REQ(matchNodes(match_node, node));
+      match->set(match_node, node);
+      node_frontier->emplace_back(std::make_pair(match_node, node));
+    } while (0);
+
+    auto match_uses = value_pair.first->uses();
+    auto uses = value_pair.second->uses();
+    for (auto i = 0; i < std::min(uses.size(), match_uses.size()); ++i) {
+      auto match_user = match_uses[i].user;
+      auto user = uses[i].user;
+      if (match->get(match_user) == user) {
+        std::stringstream ss;
+        ss << "[info] already matched node " << match_user << " to " << user;
+        log(ss.str());
+        continue;
+      }
+      REQ(user->owningBlock() == current_block_);
+      REQ(matchNodes(match_user, user));
+      node_frontier->emplace_back(std::make_pair(match_user, user));
+    }
+  }
+  return true;
+}
+
+#undef REQ
+
+bool MatchIterator::resetMatchParams(
+    MatchIterator::Match* match,
+    NodeFrontier* node_frontier,
+    ValueFrontier* value_frontier) {
+  log("[info] clearing match params");
+  match->clear();
+  node_frontier->clear();
+  value_frontier->clear();
+  auto new_node = findNewNode();
+  if (!new_node) {
+    return true;
+  } else {
+    match->set(anchor_, new_node);
+  }
+  node_frontier->emplace_back(std::make_pair(anchor_, new_node));
+  return false;
+}
+
+MatchIterator::Match MatchIterator::getMatch() {
+  MatchIterator::Match match;
+  NodeFrontier node_frontier;
+  ValueFrontier value_frontier;
+
+  if (resetMatchParams(&match, &node_frontier, &value_frontier)) {
+    return match;
+  }
+
+  // Exclude Param and Return nodes from the match graph.
+  while (match.size() < match_graph_.size() - 2) {
+    // If we have a frontier, continue matching
+    auto frontier_match =
```

Comment:
Why not `bool` instead of `auto`? I thought the returned value is some pointer when I first read this :)